zkvm/src/predicate.rs

```diff
@@ -181,75 +151,195 @@ impl Into<CompressedRistretto> for Predicate {
+        let precomputed_point = {
+            let h = adjustment_factor;
+            let x = key.0.decompress().ok_or(VMError::InvalidPoint)?;
+            let p = x + h * PedersenGens::default().B;
+            p.compress()
+        };
 
-    fn bytecode(prog: &Program) -> Vec<u8> {
-        let mut prog_vec = Vec::new();
-        prog.encode(&mut prog_vec);
-        prog_vec
+        Ok(Self {
+            leaves,
+            blinding_key,
+            key,
+            precomputed_point,
+            adjustment_factor,
+        })
     }
 
-    #[test]
-    fn valid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let pred = Predicate::unblinded_program(prog.clone());
-        let op = pred.prove_program_predicate(&bytecode(&prog), &[]);
-        assert!(op.verify().is_ok());
+    /// Returns the adjustment factor for signing
+    pub fn adjustment_factor(&self) -> Scalar {
+        self.adjustment_factor
     }
 
-    #[test]
-    fn invalid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let prog2 = Program::build(|p| p.dup(1));
-        let pred = Predicate::unblinded_program(prog);
-        let op = pred.prove_program_predicate(&bytecode(&prog2), &[]);
-        assert!(op.verify().is_err());
+    /// Creates the call proof and returns that with the program at an index.
+    pub fn create_callproof_program(
+        &self,
+        prog_index: usize,
+    ) -> Result<(CallProof, Program), VMError> {
+        let poss_leaf = &self.leaves[2 * prog_index];
+        let leaf_index = match poss_leaf {
+            PredicateLeaf::Blinding(_) => 2 * prog_index + 1,
+            PredicateLeaf::Program(_) => 2 * prog_index,
+        };
+        let tree = MerkleTree::build(b"ZkVM.taproot", &self.leaves);
+        let neighbors = tree.create_path(leaf_index).unwrap();
+        let call_proof = CallProof {
+            verification_key: self.key,
+            neighbors: neighbors,
+        };
+        let program = self.leaves[leaf_index].clone().to_program()?;
+        Ok((call_proof, program))
     }
 
-    #[test]
-    fn valid_disjunction1() {
-        let gens = PedersenGens::default();
+    fn commit_leaves(progs: &Vec<Program>, blinding_key: [u8; 32]) -> Vec<PredicateLeaf> {
+        let mut t = Transcript::new(b"ZkVM.taproot-derive-blinding");
+        let n: u64 = progs.len() as u64;
+        t.commit_u64(b"n", n);
+        t.commit_bytes(b"key", &blinding_key);
+        for prog in progs.iter() {
+            let mut buf = Vec::new();
+            prog.encode(&mut buf);
+            t.commit_bytes(b"prog", &buf);
+        }
+
+        let mut leaves = Vec::new();
+        for prog in progs.iter() {
+            let blinding = t.challenge_scalar(b"blinding");
+            let blinding_leaf = PredicateLeaf::Blinding(blinding.to_bytes());
+            let program_leaf = PredicateLeaf::Program(prog.clone());
+            if blinding.to_bytes()[31] & 1 == 0 {
```

Comment:
i'd use the lowest bit to keep it somewhat simpler (not that it matters - also if blinding was a scalar, top 4 bits would be zero, so choosing 8th-from-the-top bit is kinda weird):

```suggestion
            if blinding[0] & 1 == 0 {
```